### ISSUE TITLE
Threads failing to spin up after IIS recycle

### ISSUE THREAD

**Issue by trace-andreason** [JeffDarchuk/LyticsSitecoreConnector#8]
We have a client reporting that after their scheduled IIS recycle, sometimes their server fails to fully spin up. I know that you no longer have access to a Lytics environment to make any changes, but I'm wondering if you could give your thoughts on my theory as someone with more Sitecore experience than I do.

This client is using a multi thread setup on their application pool. Here is an example of logs that they see when the error occurs:

```
ManagedPoolThread #16 06:11:22 INFO  Job started: LyticsSitecoreConnector.Agent.RulesRefreshAgent
ManagedPoolThread #16 06:11:22 INFO  Job ended: LyticsSitecoreConnector.Agent.RulesRefreshAgent (units processed: )
28024 06:11:23 INFO  HttpModule is being initialized
7780 06:11:24 INFO  HttpModule is being initialized
27560 06:11:26 INFO  HttpModule is being initialized
45108 06:11:28 INFO  HttpModule is being initialized
34316 06:11:31 INFO  HttpModule is being initialized
23688 06:11:33 INFO  HttpModule is being initialized
39892 06:11:35 INFO  HttpModule is being initialized
35168 06:11:37 ERROR Application error.
Exception: System.Web.HttpException
Message: Request timed out.

32076 06:11:37 ERROR Application error.
Exception: System.Web.HttpException
Message: Request timed out.

39064 06:11:37 ERROR Application error.
Exception: System.Web.HttpException
Message: Request timed out.

19080 06:11:37 ERROR Application error.
Exception: System.Web.HttpException
Message: Request timed out.
```

What I think is happening is that when the IIS recycle happens, the initialize pipeline gets run for each thread and all of these threads are now competing to open the LyticsSitecoreConnector.Lytics.zip file in the packages folder. It looks like they have quite a few threads running just based on the logs showing the 16th thread. I started to focus in on this infinite while loop checking if a file is locked:
https://github.com/JeffDarchuk/LyticsSitecoreConnector/blob/master/LyticsSitecoreConnector/Pipeline/Initialize/InitializeLytics.cs#L95-L123
I noticed that in your token manager project, you bound this while loop to 15 iterations:
https://github.com/JeffDarchuk/SCTokenManager/blob/master/Source/TokenManager/Pipelines/Initialize/InitializeTokenManager.cs#L92-L118

I looked through the commit history of the project, but didn't see any specific references as to why this change was added. Do you know if that while loop cap was added because of an issue like this?

I'm also wondering about the IsLocked function here: 
https://github.com/JeffDarchuk/LyticsSitecoreConnector/blob/master/LyticsSitecoreConnector/Pipeline/Initialize/InitializeLytics.cs#L154-L177

Because the file is closed after the function is called, wouldn't that defeat the purpose of checking if the file is closed in the first place? After running that function, it seems that you couldn't guarantee that the file is actually not opened somewhere else which makes me worried about some potential deadlock in a multi threaded setup.

Your thoughts on this would be greatly appreciated!

**Comment by JeffDarchuk** [comment]
The 15 iteration loop was added to prevent things like this, it should definitely be in the lytics project too.  The error messages you're getting are HTTP timeouts and don't give much to go on, but not a direct file lock error.

What is this multi-threaded startup?  I've never heard of such a thing before.  I would suspect that such a process would have many issues all across Sitecore.  

Does disabling the Lytics connector all together prevent the issue entirely?

**Comment by trace-andreason** [comment]
Maybe I'm not using the correct terminology. I mean that they are running multiple worker threads in their application pool all serving the same site. I guess IIS would call that a "web garden"? I'd actually have to confirm this but just based on the thread numbers in the logs I was assuming it was the case.

Here is the rest of the problem description we got from the client:

```
This is not a consistent thing… but at various points we would have one of our three webservers run into this problem each morning for two or three days at a time. What really got our attention was when two of our production servers went down in the middle of the day with the same pattern in the logs. That happened this month and drove us to resolve the issue.

I was not involved with setting up the connector and mistakenly thought we only needed it on in the Content Management (CM) environment. So I proposed we simply remove the configuration for it in Content Delivery (CD). After speaking to Vivek and content editors it looks like the connector won’t work unless its configured properly in CD. But we are worried about turning it on again. Thus far it looks like removing the LyticsSitecoreConnector config in CD has solved our problem.

The fact that this is intermittent and typically only affecting one webserver makes this a hard one to diagnose. Hope you folks can point us in the right direction! (One important note – we are on Sitecore 8.1 with a few hand-applied patches. So if you know about some patching that made it into 8.2 it’s possible we don’t have that yet).
```

What jumps out at me here is that if they have two environments, and they remove the config from one environment, then everything seems fine. I'm assuming that their CD environments spans two web servers and their CM environment is just a single web server. I assume that with no Lytics config in the CD environment, the initialize pipeline would not be run on an IIS recycle for 2/3 of the environments. That would mean you have a single environment running the initialize pipeline and the issue goes away. Thoughts?

**Comment by JeffDarchuk** [comment]
all IIS environments run using a thread pool, however the initialize pipeline runs in a serial manner on a single thread from the pool, so internally in the environment there wouldn't be thread contention to worry about.  Having two separate CD servers wouldn't do this either as they would have their own backing filesystem.  I suspect you may be right on your initial suspicion, that the while loop is never breaking because the file is never unlocking. 

Truth be told, i think that whole block doesn't need to be run on the CD boxes at all: 
			AssessSitecoreState();
			ActivateLyticsRules();

These should only be ran on the CM box as the content items will get out to the CD database through normal SC publishes.

**Comment by trace-andreason** [comment]
What do you think would be a good fix in this case in addition to the while loop cap? Could a flag be added to the config to disable running those two functions during initialization?

I'm still a little confused as to why the file is becoming locked in the first place. Is it because the file isn't unlocked after the initialize pipeline is run the first time? So on subsequent initializations the file would still be in a locked state.

I guess looking at the IsLocked function, it catches all IO Exceptions, so its not necessarily locked.

From the C# docs it could be on of:
- DirectoryNotFoundException
- EndOfStreamException
- FileNotFoundException
- FileLoadException
- PathTooLongException

It definitely would be a good idea to bound this check because if any of these exceptions were being thrown, you'd definitely never be getting out of that loop.

**Comment by JeffDarchuk** [comment]
yeah, i think adding the loop cap is a good idea as well as either adding a flag as you suggested or attempting to find a master database in the code and if no master database is found assume we're running on a CD environment and not run those two methods.

I've never seen a file lock error in this way before, it's certainly possible that code could be written in a smarter way.

Ultimately i can't do any development on this project due to not having an environment myself anymore.  However if you want to make changes and submit a PR i can update the nuget package.  Alternatively, feel free to incorporate the code into your code base directly with the fixes, either way is ok with me.
